mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/MapperResultProducer.java

```diff
@@ -0,0 +1,107 @@
+ *   &#64;Select
+ *   CustomFuture&lt;Product&gt; findById(UUID productId);
+ * }
+ * </pre>
+ *
+ * See the javadocs of the methods in this interface for more explanations.
+ */
+public interface MapperResultProducer {
+
+  /**
+   * Checks if this producer can handle a particular result type.
+   *
+   * <p>This will be invoked at runtime to select a producer: if a DAO method declares a return type
+   * that is not supported natively, then the mapper generates an implementation which, for every
+   * invocation, iterates through all the producers <em>in the order that they were registered</em>,
+   * and picks the first one where {@code canProduce()} returns true.
+   *
+   * @param resultType the DAO method's declared return type. If checking the top-level type is
+   *     sufficient, then {@link GenericType#getRawType()} should do the trick. If you need to
+   *     recurse into the type arguments, call {@link GenericType#getType()} and use the {@code
+   *     java.lang.reflect} APIs.
+   */
+  boolean canProduce(GenericType<?> resultType);
+
+  /**
+   * Executes the statement generated by the mapper, and converts the result to the expected type.
+   *
+   * <p>This will be executed at runtime, every time the DAO method is called.
+   *
+   * @param statement the statement, ready to execute: the mapper has already bound all the values,
+   *     and set all the necessary attributes (consistency, page size, etc).
+   * @param context the context in which the DAO method is executed. In particular, this is how you
+   *     get access to the {@linkplain MapperContext#getSession() session}.
+   * @param entityHelper if the type to produce contains a mapped entity (e.g. {@code
+   *     ListenableFuture<Product>}), an instance of the helper class to manipulate that entity. In
+   *     particular, {@link EntityHelper#get(GettableByName) entityHelper.get()} allows you to
+   *     convert rows into entity instances. If the type to produce does not contain an entity, this
+   *     will be {@code null}.
+   * @return the object to return from the DAO method. This must match the type that this producer
+   *     was selected for, there will be an unchecked cast at runtime.
+   */
+  @SuppressWarnings("TypeParameterUnusedInFormals")
+  <EntityT> Object execute(
+      Statement<?> statement, MapperContext context, EntityHelper<EntityT> entityHelper);
+
+  /**
+   * Surfaces any error encountered in the DAO method (either in the generated mapper code that
+   * builds the statement, or during invocation of {@link #execute}).
+   *
+   * <p>For some result types, it is expected that errors will be wrapped in some sort of container
+   * instead of thrown directly; for example a failed future or publisher.
+   *
+   * <p>If rethrowing is the right thing to do, then it is perfectly fine to do so from this method
+   * (note however that you'll need to wrap checked exceptions).
+   */
+  Object wrapError(Throwable error);
```

Comment:
Wouldn't it be a bit more user-friendly if this method declared to throw `Exception`? We could then catch it in call sites and convert the thrown error into something more appropriate.